File: backend/taskflow-laravel/app/Http/Controllers/StatsController.php
<?php

namespace App\Http\Controllers;

use App\Http\Traits\ApiResponseTrait;
use App\Models\Proyecto;
use App\Models\Tarea;
use App\Models\Usuario;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

class StatsController extends Controller
{
  use ApiResponseTrait;

  /**
   * Estadísticas y datos para gráficos sobre proyectos.
   * Acceso: ADMIN only.
   */
  public function proyectos(Request $request)
  {
    $actor = $request->user();
    if (! $actor || ! method_exists($actor, 'hasRole') || ! $actor->hasRole('ADMIN')) {
      return $this->errorResponse('No autorizado', 403);
    }

    // Totales generales
    $totalProjects = Proyecto::count();
    $totalTasks = Tarea::count();

    $tasksByEstado = Tarea::select('estado', DB::raw('count(*) as count'))
      ->groupBy('estado')
      ->get()
      ->pluck('count', 'estado')
      ->toArray();

    $completed = $tasksByEstado['COMPLETADA'] ?? 0;
    $percentCompletedOverall = $totalTasks > 0 ? round($completed / $totalTasks * 100, 2) : 0;

    // Projects detailed: for charts (top projects by completion, tasks per project)
    $projects = Proyecto::withCount(['tareas as total_tareas', 'usuarios as assigned_users_count'])->get();

    $projectsData = $projects->map(function ($p) {
      $counts = Tarea::where('proyecto_id', $p->id)
        ->select('estado', DB::raw('count(*) as c'))
        ->groupBy('estado')
        ->pluck('c', 'estado')
        ->toArray();

      $total = $p->total_tareas ?? 0;
      $completed = $counts['COMPLETADA'] ?? 0;
      $percent = $total > 0 ? round($completed / $total * 100, 2) : 0;

      return [
        'id' => $p->id,
        'titulo' => $p->titulo,
        'total_tareas' => $total,
        'tasks_by_estado' => $counts,
        'percent_completed' => $percent,
        'assigned_users_count' => $p->assigned_users_count ?? 0,
        'fecha_creacion' => $p->fecha_creacion,
      ];
    })->values();

    // Top 5 proyectos por porcentaje completado (con al menos 1 tarea)
    $topProjects = $projectsData->filter(function ($p) {
      return $p['total_tareas'] > 0;
    })->sortByDesc('percent_completed')->take(5)->values();

    // Tasks per month (últimos 6 meses)
    $tasksPerMonth = Tarea::select(DB::raw("DATE_FORMAT(fecha_creacion, '%Y-%m') as ym"), DB::raw('count(*) as c'))
      ->where('fecha_creacion', '>=', now()->subMonths(6))
      ->groupBy('ym')
      ->orderBy('ym')
      ->get()
      ->mapWithKeys(function ($row) {
        return [$row->ym => (int) $row->c];
      });

    return $this->successResponse('Estadísticas de proyectos', [
      'total_projects' => $totalProjects,
      'total_tasks' => $totalTasks,
      'tasks_by_estado' => $tasksByEstado,
      'percent_completed_overall' => $percentCompletedOverall,
      'projects' => $projectsData,
      'top_projects_by_completion' => $topProjects,
      'tasks_per_month' => $tasksPerMonth,
    ]);
  }

  /**
   * Estadísticas y datos para gráficos sobre tareas y rendimiento por usuario.
   * Acceso: ADMIN only.
   */
  public function tareas(Request $request)
  {
    $actor = $request->user();
    if (! $actor || ! method_exists($actor, 'hasRole') || ! $actor->hasRole('ADMIN')) {
      return $this->errorResponse('No autorizado', 403);
    }

    $totalTasks = Tarea::count();
    $tasksByEstado = Tarea::select('estado', DB::raw('count(*) as count'))
      ->groupBy('estado')
      ->get()
      ->pluck('count', 'estado')
      ->toArray();

    $overdue = Tarea::whereNotNull('fecha_vencimiento')
      ->where('fecha_vencimiento', '<', now()->toDateString())
      ->where('estado', '!=', 'COMPLETADA')
      ->count();

    // Rendimiento por usuario: assigned, completadas, porcentaje
    $users = Usuario::where('activo', 1)->get();

    $userPerf = $users->map(function ($u) {
      $assigned = Tarea::where('asignado_a', $u->id)->count();
      $completed = Tarea::where('asignado_a', $u->id)->where('estado', 'COMPLETADA')->count();
      $overdue = Tarea::where('asignado_a', $u->id)
        ->whereNotNull('fecha_vencimiento')
        ->where('fecha_vencimiento', '<', now()->toDateString())
        ->where('estado', '!=', 'COMPLETADA')
        ->count();

      $percent = $assigned > 0 ? round($completed / $assigned * 100, 2) : 0;

      return [
        'usuario_id' => $u->id,
        'nombre' => $u->nombre,
        'assigned_count' => $assigned,
        'completed_count' => $completed,
        'overdue_count' => $overdue,
        'completion_rate' => $percent,
      ];
    })->sortByDesc('assigned_count')->values();

    // Top performers (por completion_rate, al menos 1 tarea asignada)
    $topPerformers = $userPerf->filter(function ($u) {
      return $u['assigned_count'] > 0;
    })->sortByDesc('completion_rate')->take(10)->values();

    return $this->successResponse('Estadísticas de tareas', [
      'total_tasks' => $totalTasks,
      'tasks_by_estado' => $tasksByEstado,
      'overdue_tasks' => $overdue,
      'user_performance' => $userPerf,
      'top_performers' => $topPerformers,
    ]);
  }
}

File: backend/api/src/main/java/com/doublevistudio/api/controller/StatsController.java
package com.doublevistudio.api.controller;

import com.doublevistudio.api.dto.ResponseWrapper;
import com.doublevistudio.api.model.Proyecto;
import com.doublevistudio.api.model.Tarea;
import com.doublevistudio.api.model.Usuario;
import com.doublevistudio.api.model.UsuarioProyecto;
import com.doublevistudio.api.model.UsuarioRol;
import com.doublevistudio.api.repository.ProyectoRepository;
import com.doublevistudio.api.repository.TareaRepository;
import com.doublevistudio.api.repository.UsuarioProyectoRepository;
import com.doublevistudio.api.repository.UsuarioRepository;
import com.doublevistudio.api.repository.UsuarioRolRepository;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.time.Instant;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/stats")
public class StatsController {

    @Autowired
    private ProyectoRepository proyectoRepository;

    @Autowired
    private TareaRepository tareaRepository;

    @Autowired
    private UsuarioProyectoRepository usuarioProyectoRepository;

    @Autowired
    private UsuarioRepository usuarioRepository;

    @Autowired
    private UsuarioRolRepository usuarioRolRepository;

    private static final DateTimeFormatter OUT_FMT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    @GetMapping(value = "/proyectos", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<ResponseWrapper<Map<String, Object>>> proyectosStats(HttpServletRequest request) {
        Object attr = request.getAttribute("currentUserId");
        if (attr == null) {
            ResponseWrapper<Map<String, Object>> w = new ResponseWrapper<>();
            w.setStatus("error");
            w.setMessage("No autorizado");
            w.setTimestamp(Instant.now());
            w.setData(null);
            return ResponseEntity.status(401).body(w);
        }
        Long currentUserId = Long.valueOf(String.valueOf(attr));

        // solo ADMIN
        boolean isAdmin = false;
        List<UsuarioRol> urs = usuarioRolRepository.findByIdUsuarioId(currentUserId);
        if (urs != null) {
            for (UsuarioRol ur : urs) {
                if (ur.getRol() != null && "ADMIN".equalsIgnoreCase(ur.getRol().getNombre())) {
                    isAdmin = true;
                    break;
                }
            }
        }
        if (!isAdmin) {
            ResponseWrapper<Map<String, Object>> w = new ResponseWrapper<>();
            w.setStatus("error");
            w.setMessage("No tienes permisos para ver estadísticas");
            w.setTimestamp(Instant.now());
            w.setData(null);
            return ResponseEntity.status(403).body(w);
        }

        List<Proyecto> proyectos = proyectoRepository.findAll();
        List<Tarea> allTareas = tareaRepository.findAll();

        // consider only usuarios activos when counting assigned users
        // build a set of active user ids
        Set<Long> activeUserIds = usuarioRepository.findAll().stream()
                .filter(u -> Boolean.TRUE.equals(u.getActivo()))
                .map(Usuario::getId)
                .collect(Collectors.toSet());

        Map<String, Integer> tasksByEstadoGlobal = new HashMap<>();
        tasksByEstadoGlobal.put("PENDIENTE", 0);
        tasksByEstadoGlobal.put("PROGRESO", 0);
        tasksByEstadoGlobal.put("COMPLETADA", 0);

        for (Tarea t : allTareas) {
            String s = t.getEstado() != null ? t.getEstado().name() : "PENDIENTE";
            tasksByEstadoGlobal.put(s, tasksByEstadoGlobal.getOrDefault(s, 0) + 1);
        }

        int totalTasks = allTareas.size();
        int totalProjects = proyectos.size();
        double percentCompletedOverall = totalTasks == 0 ? 0 : (tasksByEstadoGlobal.getOrDefault("COMPLETADA", 0) * 100.0) / totalTasks;
        percentCompletedOverall = Math.round(percentCompletedOverall * 100.0) / 100.0;

        List<Map<String, Object>> projectsList = new ArrayList<>();

        for (Proyecto p : proyectos) {
            Map<String, Object> projMap = new HashMap<>();
            projMap.put("id", p.getId());
            projMap.put("titulo", p.getTitulo());

            // tareas del proyecto
            List<Tarea> tareasProyecto = allTareas.stream()
                    .filter(t -> t.getProyecto() != null && Objects.equals(t.getProyecto().getId(), p.getId()))
                    .collect(Collectors.toList());

            Map<String, Integer> byEstado = new HashMap<>();
            for (Tarea t : tareasProyecto) {
                String s = t.getEstado() != null ? t.getEstado().name() : "PENDIENTE";
                byEstado.put(s, byEstado.getOrDefault(s, 0) + 1);
            }

            int totalTareasProyecto = tareasProyecto.size();
            double percentCompleted = totalTareasProyecto == 0 ? 0 : (byEstado.getOrDefault("COMPLETADA", 0) * 100.0) / totalTareasProyecto;
            percentCompleted = Math.round(percentCompleted * 100.0) / 100.0;

            // assigned users count (only activos)
            List<UsuarioProyecto> ups = usuarioProyectoRepository.findByIdProyectoId(p.getId());
            long assignedActiveCount = ups.stream()
                    .map(UsuarioProyecto::getUsuario)
                    .filter(Objects::nonNull)
                    .map(Usuario::getId)
                    .filter(activeUserIds::contains)
                    .distinct()
                    .count();

            projMap.put("total_tareas", totalTareasProyecto);
            projMap.put("tasks_by_estado", byEstado);
            projMap.put("percent_completed", percentCompleted);
            projMap.put("assigned_users_count", assignedActiveCount);
            projMap.put("fecha_creacion", p.getFechaCreacion() != null ? p.getFechaCreacion().format(OUT_FMT) : null);

            projectsList.add(projMap);
        }

        // top projects by completion (descending percent_completed)
        List<Map<String, Object>> topProjects = projectsList.stream()
                .sorted((a, b) -> Double.compare((Double) b.get("percent_completed"), (Double) a.get("percent_completed")))
                .collect(Collectors.toList());

        // tasks per month (based on tarea.fechaCreacion month)
        Map<String, Integer> tasksPerMonth = new HashMap<>();
        DateTimeFormatter monthFmt = DateTimeFormatter.ofPattern("yyyy-MM");
        for (Tarea t : allTareas) {
            if (t.getFechaCreacion() != null) {
                String key = t.getFechaCreacion().format(monthFmt);
                tasksPerMonth.put(key, tasksPerMonth.getOrDefault(key, 0) + 1);
            }
        }

        Map<String, Object> data = new HashMap<>();
        data.put("total_projects", totalProjects);
        data.put("total_tasks", totalTasks);
        data.put("tasks_by_estado", tasksByEstadoGlobal);
        data.put("percent_completed_overall", percentCompletedOverall);
        data.put("projects", projectsList);
        data.put("top_projects_by_completion", topProjects);
        data.put("tasks_per_month", tasksPerMonth);

        ResponseWrapper<Map<String, Object>> wrap = new ResponseWrapper<>();
        wrap.setStatus("success");
        wrap.setMessage("Estadísticas de proyectos");
        wrap.setTimestamp(Instant.now());
        wrap.setData(data);
        return ResponseEntity.ok(wrap);
    }

    @GetMapping(value = "/tareas", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<ResponseWrapper<Map<String, Object>>> tareasStats(HttpServletRequest request) {
        Object attr = request.getAttribute("currentUserId");
        if (attr == null) {
            ResponseWrapper<Map<String, Object>> w = new ResponseWrapper<>();
            w.setStatus("error");
            w.setMessage("No autorizado");
            w.setTimestamp(Instant.now());
            w.setData(null);
            return ResponseEntity.status(401).body(w);
        }
        Long currentUserId = Long.valueOf(String.valueOf(attr));

        // solo ADMIN
        boolean isAdmin = false;
        List<UsuarioRol> urs = usuarioRolRepository.findByIdUsuarioId(currentUserId);
        if (urs != null) {
            for (UsuarioRol ur : urs) {
                if (ur.getRol() != null && "ADMIN".equalsIgnoreCase(ur.getRol().getNombre())) {
                    isAdmin = true;
                    break;
                }
            }
        }
        if (!isAdmin) {
            ResponseWrapper<Map<String, Object>> w = new ResponseWrapper<>();
            w.setStatus("error");
            w.setMessage("No tienes permisos para ver estadísticas");
            w.setTimestamp(Instant.now());
            w.setData(null);
            return ResponseEntity.status(403).body(w);
        }

        List<Tarea> allTareas = tareaRepository.findAll();

        Map<String, Integer> tasksByEstado = new HashMap<>();
        tasksByEstado.put("PENDIENTE", 0);
        tasksByEstado.put("PROGRESO", 0);
        tasksByEstado.put("COMPLETADA", 0);

        int overdueTasks = 0;
        LocalDate today = LocalDate.now();
        for (Tarea t : allTareas) {
            String s = t.getEstado() != null ? t.getEstado().name() : "PENDIENTE";
            tasksByEstado.put(s, tasksByEstado.getOrDefault(s, 0) + 1);

            if (t.getFechaVencimiento() != null) {
                if (t.getFechaVencimiento().isBefore(today) && (t.getEstado() == null || !"COMPLETADA".equals(t.getEstado().name()))) {
                    overdueTasks++;
                }
            }
        }

        // performance por usuario (solo usuarios activos)
        // Incluimos solo usuarios activos que además tengan el rol USER (no ADMIN)
        List<Usuario> activeUsers = usuarioRepository.findAll().stream()
                .filter(u -> Boolean.TRUE.equals(u.getActivo()))
                .filter(u -> {
                    List<UsuarioRol> rolesForUser = usuarioRolRepository.findByIdUsuarioId(u.getId());
                    if (rolesForUser == null) return false;
                    return rolesForUser.stream().anyMatch(r -> r.getRol() != null && "USER".equalsIgnoreCase(r.getRol().getNombre()));
                })
                .collect(Collectors.toList());

        // init stats map
        Map<Long, Map<String, Object>> perfMap = new HashMap<>();
        for (Usuario u : activeUsers) {
            Map<String, Object> m = new HashMap<>();
            m.put("usuario_id", u.getId());
            m.put("nombre", u.getNombre());
            m.put("assigned_count", 0);
            m.put("completed_count", 0);
            m.put("overdue_count", 0);
            m.put("completion_rate", 0.0);
            perfMap.put(u.getId(), m);
        }

        // aggregate tasks per assigned user (only active users counted)
        for (Tarea t : allTareas) {
            if (t.getAsignadoA() == null) continue;
            Long uid = t.getAsignadoA().getId();
            if (!perfMap.containsKey(uid)) continue; // skip inactive users
            Map<String, Object> m = perfMap.get(uid);
            int assigned = (int) m.get("assigned_count");
            assigned++;
            m.put("assigned_count", assigned);
            if (t.getEstado() != null && "COMPLETADA".equals(t.getEstado().name())) {
                int comp = (int) m.get("completed_count");
                m.put("completed_count", comp + 1);
            }
            if (t.getFechaVencimiento() != null && t.getFechaVencimiento().isBefore(today)
                    && (t.getEstado() == null || !"COMPLETADA".equals(t.getEstado().name()))) {
                int ov = (int) m.get("overdue_count");
                m.put("overdue_count", ov + 1);
            }
        }

        // compute completion_rate and build list
        List<Map<String, Object>> userPerformance = new ArrayList<>();
        for (Map.Entry<Long, Map<String, Object>> e : perfMap.entrySet()) {
            Map<String, Object> m = e.getValue();
            int assigned = (int) m.get("assigned_count");
            int completed = (int) m.get("completed_count");
            double rate = assigned == 0 ? 0.0 : (completed * 100.0) / assigned;
            rate = Math.round(rate * 100.0) / 100.0;
            m.put("completion_rate", rate);
            userPerformance.add(m);
        }

        // top performers: users with assigned_count>0 ordered by completion_rate desc then completed_count desc
        List<Map<String, Object>> topPerformers = userPerformance.stream()
                .filter(m -> (int) m.get("assigned_count") > 0)
                .sorted((a, b) -> {
                    int cmp = Double.compare((Double) b.get("completion_rate"), (Double) a.get("completion_rate"));
                    if (cmp != 0) return cmp;
                    int cb = (int) b.get("completed_count");
                    int ca = (int) a.get("completed_count");
                    return Integer.compare(cb, ca);
                })
                .limit(5)
                .collect(Collectors.toList());

        Map<String, Object> data = new HashMap<>();
        data.put("total_tasks", allTareas.size());
        data.put("tasks_by_estado", tasksByEstado);
        data.put("overdue_tasks", overdueTasks);
        data.put("user_performance", userPerformance);
        data.put("top_performers", topPerformers);

        ResponseWrapper<Map<String, Object>> wrap = new ResponseWrapper<>();
        wrap.setStatus("success");
        wrap.setMessage("Estadísticas de tareas");
        wrap.setTimestamp(Instant.now());
        wrap.setData(data);
        return ResponseEntity.ok(wrap);
    }
}
